File: rust/crates/mass-spec/tests/nta_interface_smoke.rs
use std::collections::BTreeSet;
use streamfind_rust_core::MethodRegistry;

#[test]
fn nta_interface_registers_catalogue_methods_without_data_files() {
    let entries = streamfind_rust_test_support::catalogue_entries();
    let expected_ids: BTreeSet<String> = entries
        .iter()
        .filter(|entry| {
            entry["kind"] == "method"
                && entry["domain"] == "mass_spec"
                && entry["canonical_id"]
                    .as_str()
                    .is_some_and(|id| id.starts_with("mass_spec."))
        })
        .map(|entry| entry["canonical_id"].as_str().unwrap().to_owned())
        .collect();
    assert_eq!(expected_ids.len(), 19, "expected 19 catalogue Methods");

    let mut methods = MethodRegistry::default();
    streamfind_rust_mass_spec::register_methods(&mut methods).unwrap();
    let registered = methods.list("mass_spec");
    let actual_ids: BTreeSet<String> = registered
        .iter()
        .filter_map(|method| method["id"].as_str())
        .filter(|id| id.starts_with("mass_spec."))
        .map(str::to_owned)
        .collect();
    assert_eq!(
        actual_ids, expected_ids,
        "registered Method set differs from catalogue"
    );

    for id in &expected_ids {
        let entry = entries
            .iter()
            .find(|entry| entry["canonical_id"].as_str() == Some(id))
            .unwrap_or_else(|| panic!("missing catalogue entry for {id}"));
        assert_eq!(entry["kind"], "method");
        assert_eq!(entry["domain"], "mass_spec");
    }

    let find_features = methods.get("mass_spec.find_features").unwrap();
    assert!(find_features.cacheable);
    assert!(find_features.single_occurrence);
    assert!(find_features.required_methods.is_empty());
}

File: cpp/tests/unit/nta_interface_smoke.cpp
#include <iostream>
#include <set>
#include <string>

#include "streamfind/catalogue.hpp"
#include "streamfind/mass_spec/register.hpp"

#ifndef STREAMFIND_CATALOGUE_PATH
#error STREAMFIND_CATALOGUE_PATH is required
#endif

int main() {
    const auto entries = streamfind::catalogue::load(STREAMFIND_CATALOGUE_PATH);
    if (!entries) {
        std::cerr << "NTA interface: catalogue load failed\n";
        return 1;
    }

    std::set<std::string> expected_ids;
    for (const auto &entry : *entries) {
        if (entry.value("kind", "") == "method" && entry.value("domain", "") == "mass_spec") {
            expected_ids.insert(entry.value("canonical_id", ""));
        }
    }
    if (expected_ids.size() != 19) {
        std::cerr << "NTA interface: expected 19 catalogue Methods, found " << expected_ids.size() << '\n';
        return 1;
    }

    streamfind::MethodRegistry methods;
    streamfind::OperationRegistry operations;
    streamfind::mass_spec::register_plugin(*entries, methods, operations);
    const auto registered = methods.list("mass_spec");
    std::set<std::string> actual_ids;
    for (const auto &definition : registered) {
        if (definition.id.rfind("mass_spec.", 0) == 0) actual_ids.insert(definition.id);
    }
    if (actual_ids != expected_ids) {
        std::cerr << "NTA interface: registered Method set differs from catalogue\n";
        return 1;
    }
    for (const auto &entry : *entries) {
        if (entry.value("kind", "") == "method" && entry.value("domain", "") == "mass_spec" &&
            entry.value("module_id", "") != "mass_spec.nta" &&
            entry.value("module_id", "") != "mass_spec.chromatograms") {
            std::cerr << "NTA interface: invalid module ownership\n";
            return 1;
        }
    }

    const auto *find_features = methods.find("mass_spec.find_features");
    if (!find_features->definition().cacheable || !find_features->definition().single_occurrence) {
        std::cerr << "NTA interface: find_features lifecycle metadata mismatch\n";
        return 1;
    }
    return 0;
}
